_posts/2022-01-24-publishing-typescript-npm-modules.md

```diff
@@ -0,0 +1,193 @@
+
+```sh
+npm install typescript @mui/material --save-dev
+```
+
+Once we have TypeScript installed, run `npx tsc init` which will create a default `tsconfig.json` file which is used to help configure TypeScript. The reason we use `npx` to run the TypeScript compiler (`tsc`) is because it will look for the command in the locally installed dependencies rather than trying to use a globally installed version which would potentially cause version mismatch issues.
+
+> If you ever need to find out what each of the configuration options means you can use this [handy guide](https://www.typescriptlang.org/tsconfig).
+
+## Adding the Theme
+
+Now, we are going to build our custom theme. Start by creating a `src` directory in the root of your project and within that we are going to create a new file called `index.ts`. In this file we'll initialise the Material UI theme and set the primary and secondary colours which will be shared among projects using this theme.
+
+> You can find out more about the [Material UI theming options here](https://mui.com/customization/theming/)
+
+```js
+import { createTheme } from '@mui/material/styles';
+
+const theme = createTheme({
+	palette: {
+		primary: {
+			main: '#40bfb4',
+		},
+		secondary: {
+			main: '#e15554',
+		},
+	},
+});
+
+export default theme;
+```
+
+That's it! That's all this project needs in the way of code, but how do we build it and where should we put the output?
+
+Well within our `tsconfig.json`, we can add the following lines to the `compilerOptions`.
+
+```json
+"compilerOptions": {
+    "outDir": "./dist",
+    "declaration": true
+}
+```
+
+The `outDir` parameter tells TypeScript where to put the compiled TS files while setting `declaration` to `true` ensures the type definition files are generated.
+
+Next, we should add a new script to the `package.json` that will define how we build our project. At this stage, you might also want to add a script called `prepublishOnly` that tells NPM to build the project prior to publishing it to the NPM registry.
+
+```json
+"scripts": {
+    "build": "tsc",
+    "prepublishOnly": "npm run build"
+}
+```
+
+Let's try running it with:
+
+```sh
+npm run build
+```
+
+Once that command completes successfully, you should notice some files have been added to your project in the `dist` file. If you look through those files, you may notice that it includes references to `@mui/material/styles` since we used that library within the theme code we wrote above. But how do we ensure that projects using our module have `@mui/material` installed?
+
+Well, that's where peer dependencies come in!
+
+Peer dependencies are a way of defining what packages our module requires and the particular versions it is compatible with. When installing our module, NPM checks if the peer dependencies exist already within the project and if not, it installs those dependencies. So let's add some `peerDependencies` to our `package.json`.
```

Comment:
Your explanation here is (i believe) not technically correct.

The peerDependency here is to tell the consumer of this package that it is compatible when used against a particular version of the consumer package.

In this case, declaring a peer dependency of @mui/material ^5.0.0 says that the consumer should take a dependency on @mui/material ^5.0.0 - they'll get a warning if they are currently using anything lower (either dev dependency or dependency), and it'll get automatically installed if they aren't currently depending on it.

I'd also include them in "dependencies" purely to indicate that it's a runtime dependency of the package.

Reply:
Yeah I thought that was the case however from https://docs.npmjs.com/cli/v8/configuring-npm/package-json#peerdependencies

> In npm versions 3 through 6, peerDependencies were not automatically installed, and would raise a warning if an invalid version of the peer dependency was found in the tree. As of npm v7, peerDependencies are installed by default.

I might add a note that this is the case.